src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -14,592 +14,549 @@ namespace Microsoft.CodeAnalysis.CSharp
 {
     internal sealed partial class LocalRewriter
     {
-        /// <summary>
-        /// The strategy of this rewrite is to do rewrite "locally".
-        /// We analyze arguments of the concat in a shallow fashion assuming that 
-        /// lowering and optimizations (including this one) is already done for the arguments.
-        /// Based on the arguments we select the most appropriate pattern for the current node.
-        /// 
-        /// NOTE: it is not guaranteed that the node that we chose will be the most optimal since we have only 
-        ///       local information - i.e. we look at the arguments, but we do not know about siblings.
-        ///       When we move to the parent, the node may be rewritten by this or some another optimization.
-        ///       
-        /// Example:
-        ///     result = ( "abc" + "def" + null ?? expr1 + "moo" + "baz" ) + expr2
-        /// 
-        /// Will rewrite into:
-        ///     result = Concat("abcdef", expr2)
-        ///     
-        /// However there will be transient nodes like  Concat(expr1 + "moo")  that will not be present in the
-        /// resulting tree.
-        ///
-        /// </summary>
-        private BoundExpression RewriteStringConcatenation(SyntaxNode syntax, BinaryOperatorKind operatorKind, BoundExpression loweredLeft, BoundExpression loweredRight, TypeSymbol type)
+        private static bool IsBinaryStringConcatenation([NotNullWhen(true)] BoundBinaryOperator? binaryOperator)
+            => binaryOperator is { OperatorKind: var kind } && IsBinaryStringConcatenation(kind);
+
+        private static bool IsBinaryStringConcatenation(BinaryOperatorKind binaryOperator)
+            => binaryOperator is BinaryOperatorKind.StringConcatenation or BinaryOperatorKind.StringAndObjectConcatenation or BinaryOperatorKind.ObjectAndStringConcatenation;
+
+        private BoundExpression VisitCompoundAssignmentStringConcatenation(BoundExpression left, BoundExpression unvisitedRight, BinaryOperatorKind operatorKind, TypeSymbol resultType, SyntaxNode syntax)
         {
-            Debug.Assert(
-                operatorKind == BinaryOperatorKind.StringConcatenation ||
-                operatorKind == BinaryOperatorKind.StringAndObjectConcatenation ||
-                operatorKind == BinaryOperatorKind.ObjectAndStringConcatenation);
+            Debug.Assert(IsBinaryStringConcatenation(operatorKind));
 
             if (_inExpressionLambda)
             {
-                return RewriteStringConcatInExpressionLambda(syntax, operatorKind, loweredLeft, loweredRight, type);
+                return RewriteStringConcatInExpressionLambda(left, unvisitedRight, operatorKind, syntax, resultType);
```

Comment:
>return RewriteStringConcatInExpressionLambda(left, unvisitedRight, operatorKind, syntax, resultType); [](http://example.com/codeflow?start=16&length=101)

Is this a "dead" code? I think assignments are not allowed in expression lambdas.

Reply:
I didn't realize that, thanks. That will simplify a bunch of code.